# INFO 370 Lab 2
*Name: Brian Park*

## Q1 Data import and summary

### 1. Download the file time-series-covid19-deaths-us.csv.bz2 from canvas (or directly from the GH page linked above) and load it into a pandas dataframe.

In [87]:
import pandas as pd
df = pd.read_csv("../data/time-series-covid19-deaths-us.csv.bz2", sep="\t")
df.head(5)

UID iso2 iso3  code3    FIPS   Admin2 Province_State Country_Region  \
0  84001001   US  USA    840  1001.0  Autauga        Alabama             US   
1  84001003   US  USA    840  1003.0  Baldwin        Alabama             US   
2  84001005   US  USA    840  1005.0  Barbour        Alabama             US   
3  84001007   US  USA    840  1007.0     Bibb        Alabama             US   
4  84001009   US  USA    840  1009.0   Blount        Alabama             US   

         Lat      Long_  ... 9/27/22  9/28/22  9/29/22  9/30/22  10/1/22  \
0  32.539527 -86.644082  ...     227      227      227      227      227   
1  30.727750 -87.722071  ...     711      711      712      712      712   
2  31.868263 -85.387129  ...     102      102      103      103      103   
3  32.996421 -87.125115  ...     107      107      107      107      107   
4  33.982109 -86.567906  ...     258      258      258      258      258   

   10/2/22  10/3/22  10/4/22  10/5/22  10/6/22  
0      227      227      227      227      228  
1      712      712      712      712      714  
2      103      103      103      103      103  
3      107      107      107      107      108  
4      258      258      258      258      258  

[5 rows x 1001 columns]

### 2. Report the number of rows and columns in the dataset.

In [88]:
df.shape

(3342, 1001)

The dataset has 3342 rows and 1001 columns.

### 3. What variables does this dataset have? Report the variable names along with the data type of each variable.

In [89]:
df.dtypes

UID          int64
iso2        object
iso3        object
code3        int64
FIPS       float64
            ...   
10/2/22      int64
10/3/22      int64
10/4/22      int64
10/5/22      int64
10/6/22      int64
Length: 1001, dtype: object

### 4. Create a sub-dataframe that contains all observations but only variables FIPS, Admin2, Province_State, 3/1/20, 3/15/21, and the most recent date 10/6/22. 

In [90]:
newDf = df[["FIPS", "Admin2", "Province_State", "3/1/20", "3/15/21", "10/6/22"]]
newDf.shape

(3342, 6)

In [91]:
newDf.head(5)

FIPS   Admin2 Province_State  3/1/20  3/15/21  10/6/22
0  1001.0  Autauga        Alabama       0       95      228
1  1003.0  Baldwin        Alabama       0      294      714
2  1005.0  Barbour        Alabama       0       53      103
3  1007.0     Bibb        Alabama       0       58      108
4  1009.0   Blount        Alabama       0      129      258

## Q2 Explore deaths in Washington

### 1. Take the subset you created above and filter results for Washington state only. Store it into a new variable.

In [92]:
waDf = newDf[newDf.Province_State == "Washington"]
waDf

FIPS        Admin2 Province_State  3/1/20  3/15/21  10/6/22
3145  53001.0         Adams     Washington       0       22       42
3146  53003.0        Asotin     Washington       0       28       76
3147  53005.0        Benton     Washington       0      209      509
3148  53007.0        Chelan     Washington       0       62      184
3149  53009.0       Clallam     Washington       0        8      176
3150  53011.0         Clark     Washington       0      239      933
3151  53013.0      Columbia     Washington       0        4       16
3152  53015.0       Cowlitz     Washington       0       71      379
3153  53017.0       Douglas     Washington       0       22       81
3154  53019.0         Ferry     Washington       0        4       30
3155  53021.0      Franklin     Washington       0      103      220
3156  53023.0      Garfield     Washington       0        4        7
3157  53025.0         Grant     Washington       0       69      228
3158  53027.0  Grays Harbor     Washington       0       54      225
3159  53029.0        Island     Washington       0       28       99
3160  53031.0     Jefferson     Washington       0        2       31
3161  53033.0          King     Washington       1     1440     3194
3162  53035.0        Kitsap     Washington       0       88      389
3163  53037.0      Kittitas     Washington       0       30       84
3164  53039.0     Klickitat     Washington       0        9       55
3165  53041.0         Lewis     Washington       0       50      278
3166  53043.0       Lincoln     Washington       0        6       29
3167  53045.0         Mason     Washington       0       25      163
3168  53047.0      Okanogan     Washington       0       36      135
3169  80053.0     Out of WA     Washington       0        0        0
3170  53049.0       Pacific     Washington       0       10       70
3171  53051.0  Pend Oreille     Washington       0        7       39
3172  53053.0        Pierce     Washington       0      586     1526
3173  53055.0      San Juan     Washington       0        0        2
3174  53057.0        Skagit     Washington       0       62      235
3175  53059.0      Skamania     Washington       0        1       15
3176  53061.0     Snohomish     Washington       0      544     1313
3177  53063.0       Spokane     Washington       0      583     1544
3178  53065.0       Stevens     Washington       0       24      158
3179  53067.0      Thurston     Washington       0       79      469
3180  90053.0    Unassigned     Washington       0        4        6
3181  53069.0     Wahkiakum     Washington       0        0        9
3182  53071.0   Walla Walla     Washington       0       64      156
3183  53073.0       Whatcom     Washington       0       85      334
3184  53075.0       Whitman     Washington       0       46       97
3185  53077.0        Yakima     Washington       0      427      839

### 2. Add a new variable, “growth”, to the WA data. It should be the growth in the cumulative number of deaths through the last year, between 10/6/22 and 3/15/2021.

In [93]:
waDf["growth"] = waDf["10/6/22"] - waDf["3/15/21"]
waDf

/tmp/ipykernel_59/2746658991.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  waDf["growth"] = waDf["10/6/22"] - waDf["3/15/21"]


FIPS        Admin2 Province_State  3/1/20  3/15/21  10/6/22  growth
3145  53001.0         Adams     Washington       0       22       42      20
3146  53003.0        Asotin     Washington       0       28       76      48
3147  53005.0        Benton     Washington       0      209      509     300
3148  53007.0        Chelan     Washington       0       62      184     122
3149  53009.0       Clallam     Washington       0        8      176     168
3150  53011.0         Clark     Washington       0      239      933     694
3151  53013.0      Columbia     Washington       0        4       16      12
3152  53015.0       Cowlitz     Washington       0       71      379     308
3153  53017.0       Douglas     Washington       0       22       81      59
3154  53019.0         Ferry     Washington       0        4       30      26
3155  53021.0      Franklin     Washington       0      103      220     117
3156  53023.0      Garfield     Washington       0        4        7       3
3157  53025.0         Grant     Washington       0       69      228     159
3158  53027.0  Grays Harbor     Washington       0       54      225     171
3159  53029.0        Island     Washington       0       28       99      71
3160  53031.0     Jefferson     Washington       0        2       31      29
3161  53033.0          King     Washington       1     1440     3194    1754
3162  53035.0        Kitsap     Washington       0       88      389     301
3163  53037.0      Kittitas     Washington       0       30       84      54
3164  53039.0     Klickitat     Washington       0        9       55      46
3165  53041.0         Lewis     Washington       0       50      278     228
3166  53043.0       Lincoln     Washington       0        6       29      23
3167  53045.0         Mason     Washington       0       25      163     138
3168  53047.0      Okanogan     Washington       0       36      135      99
3169  80053.0     Out of WA     Washington       0        0        0       0
3170  53049.0       Pacific     Washington       0       10       70      60
3171  53051.0  Pend Oreille     Washington       0        7       39      32
3172  53053.0        Pierce     Washington       0      586     1526     940
3173  53055.0      San Juan     Washington       0        0        2       2
3174  53057.0        Skagit     Washington       0       62      235     173
3175  53059.0      Skamania     Washington       0        1       15      14
3176  53061.0     Snohomish     Washington       0      544     1313     769
3177  53063.0       Spokane     Washington       0      583     1544     961
3178  53065.0       Stevens     Washington       0       24      158     134
3179  53067.0      Thurston     Washington       0       79      469     390
3180  90053.0    Unassigned     Washington       0        4        6       2
3181  53069.0     Wahkiakum     Washington       0        0        9       9
3182  53071.0   Walla Walla     Washington       0       64      156      92
3183  53073.0       Whatcom     Washington       0       85      334     249
3184  53075.0       Whitman     Washington       0       46       97      51
3185  53077.0        Yakima     Washington       0      427      839     412

### 3. You may notice a warning: “A value is trying to be set on a copy of a slice...”. Why do we get this warning? How can we get rid of it? Repeat the previous steps in a way that you do not get the warning!

The reason why we get this warning is because the command line above returns the view of dataframe that may cause problem when filtering the data later. To get rid of this message, we need to put ".copy()" at the end of the command line to create unique dataframe.

In [94]:
waDf = newDf[newDf.Province_State == "Washington"].copy()
waDf["growth"] = waDf["10/6/22"] - waDf["3/15/21"]
waDf

FIPS        Admin2 Province_State  3/1/20  3/15/21  10/6/22  growth
3145  53001.0         Adams     Washington       0       22       42      20
3146  53003.0        Asotin     Washington       0       28       76      48
3147  53005.0        Benton     Washington       0      209      509     300
3148  53007.0        Chelan     Washington       0       62      184     122
3149  53009.0       Clallam     Washington       0        8      176     168
3150  53011.0         Clark     Washington       0      239      933     694
3151  53013.0      Columbia     Washington       0        4       16      12
3152  53015.0       Cowlitz     Washington       0       71      379     308
3153  53017.0       Douglas     Washington       0       22       81      59
3154  53019.0         Ferry     Washington       0        4       30      26
3155  53021.0      Franklin     Washington       0      103      220     117
3156  53023.0      Garfield     Washington       0        4        7       3
3157  53025.0         Grant     Washington       0       69      228     159
3158  53027.0  Grays Harbor     Washington       0       54      225     171
3159  53029.0        Island     Washington       0       28       99      71
3160  53031.0     Jefferson     Washington       0        2       31      29
3161  53033.0          King     Washington       1     1440     3194    1754
3162  53035.0        Kitsap     Washington       0       88      389     301
3163  53037.0      Kittitas     Washington       0       30       84      54
3164  53039.0     Klickitat     Washington       0        9       55      46
3165  53041.0         Lewis     Washington       0       50      278     228
3166  53043.0       Lincoln     Washington       0        6       29      23
3167  53045.0         Mason     Washington       0       25      163     138
3168  53047.0      Okanogan     Washington       0       36      135      99
3169  80053.0     Out of WA     Washington       0        0        0       0
3170  53049.0       Pacific     Washington       0       10       70      60
3171  53051.0  Pend Oreille     Washington       0        7       39      32
3172  53053.0        Pierce     Washington       0      586     1526     940
3173  53055.0      San Juan     Washington       0        0        2       2
3174  53057.0        Skagit     Washington       0       62      235     173
3175  53059.0      Skamania     Washington       0        1       15      14
3176  53061.0     Snohomish     Washington       0      544     1313     769
3177  53063.0       Spokane     Washington       0      583     1544     961
3178  53065.0       Stevens     Washington       0       24      158     134
3179  53067.0      Thurston     Washington       0       79      469     390
3180  90053.0    Unassigned     Washington       0        4        6       2
3181  53069.0     Wahkiakum     Washington       0        0        9       9
3182  53071.0   Walla Walla     Washington       0       64      156      92
3183  53073.0       Whatcom     Washington       0       85      334     249
3184  53075.0       Whitman     Washington       0       46       97      51
3185  53077.0        Yakima     Washington       0      427      839     412

### 4. Extract the newly created variable “growth” as series. Show that it is a series!

In [95]:
onlyGrowth = waDf["growth"]
onlyGrowth

3145      20
3146      48
3147     300
3148     122
3149     168
3150     694
3151      12
3152     308
3153      59
3154      26
3155     117
3156       3
3157     159
3158     171
3159      71
3160      29
3161    1754
3162     301
3163      54
3164      46
3165     228
3166      23
3167     138
3168      99
3169       0
3170      60
3171      32
3172     940
3173       2
3174     173
3175      14
3176     769
3177     961
3178     134
3179     390
3180       2
3181       9
3182      92
3183     249
3184      51
3185     412
Name: growth, dtype: int64

In [96]:
type(onlyGrowth)

pandas.core.series.Series

### 5. What is index of the growth series? What do you think, is it a useful index? What might be a better index? Why?

In [97]:
waDf["growth"].index

Int64Index([3145, 3146, 3147, 3148, 3149, 3150, 3151, 3152, 3153, 3154, 3155,
            3156, 3157, 3158, 3159, 3160, 3161, 3162, 3163, 3164, 3165, 3166,
            3167, 3168, 3169, 3170, 3171, 3172, 3173, 3174, 3175, 3176, 3177,
            3178, 3179, 3180, 3181, 3182, 3183, 3184, 3185],
           dtype='int64')

The index is a row number from the original dataframe which I don't think it's a useful index at all. Rather, we should set it up with the county name or "Admin2" to show the cumulative growth of deaths by counties.

### 6. Replace the original index with one you think is better. Show that it worked (you may just want to print a few lines).

In [98]:
onlyGrowth.index = waDf.Admin2
onlyGrowth.head(10)

Admin2
Adams        20
Asotin       48
Benton      300
Chelan      122
Clallam     168
Clark       694
Columbia     12
Cowlitz     308
Douglas      59
Ferry        26
Name: growth, dtype: int64

## Q3 Data interpretation

### 1. Sort your series of growth numbers starting with the counties with the largest growths down to the smalles. Which 10 counties did experience most growth through 2021-2022?

In [99]:
onlyGrowth.sort_values(ascending = False)

Admin2
King            1754
Spokane          961
Pierce           940
Snohomish        769
Clark            694
Yakima           412
Thurston         390
Cowlitz          308
Kitsap           301
Benton           300
Whatcom          249
Lewis            228
Skagit           173
Grays Harbor     171
Clallam          168
Grant            159
Mason            138
Stevens          134
Chelan           122
Franklin         117
Okanogan          99
Walla Walla       92
Island            71
Pacific           60
Douglas           59
Kittitas          54
Whitman           51
Asotin            48
Klickitat         46
Pend Oreille      32
Jefferson         29
Ferry             26
Lincoln           23
Adams             20
Skamania          14
Columbia          12
Wahkiakum          9
Garfield           3
San Juan           2
Unassigned         2
Out of WA          0
Name: growth, dtype: int64

King, Spokane, Pierce, Snohomish, Clark, Yakima, Thurston, Cowlitz, Kitsap, Benton counties are the top 10 counties that experienced most growth through 2021-2022.

### 2. If you did the previous question correctly, you found that King county has the largest growth (1420). What does this tell you about the efficacy of social distancing measures in Seattle?

I assume that social distancing in King country was the least effective to prevent people's deaths from COVID-19 compared to any other county in Washington. King county is where the most populations are located which may cause to undermine the efficacy of social distancing.